pytorch_lightning/metrics/metric.py

```diff
@@ -0,0 +1,220 @@
+        return super().float()
+
+    def double(self) -> torch.nn.Module:
+        """Casts all floating point parameters and buffers to ``double`` datatype.
+
+        Returns:
+            Module: self
+        """
+        self._dtype = torch.double
+        return super().double()
+
+    def half(self) -> torch.nn.Module:
+        """Casts all floating point parameters and buffers to ``half`` datatype.
+
+        Returns:
+            Module: self
+        """
+        self._dtype = torch.half
+        return super().half()
+
+
+class TensorMetric(Metric):
+    def __init__(self, name: str,
+                 reduce_group: Optional[Any] = torch.distributed.group.WORLD,
+                 reduce_op: Optional[Any] = torch.distributed.ReduceOp.SUM):
+        """
+        Base class for metric implementation operating directly on tensors.
+        All inputs and outputs will be casted to tensors if necessary.
+        Already handles DDP sync and input/output conversions
+
+        Args:
+            name: the metric's name
+            reduce_group: the process group for DDP reduces (only needed for DDP training).
+                Defaults to all processes (world)
+            reduce_op: the operation to perform during reduction within DDP (only needed for DDP training).
+                Defaults to sum.
+        """
+        super().__init__(name)
+        self._orig_call = tensor_metric(group=reduce_group,
+                                        reduce_op=reduce_op)(super().__call__)
+
+    def __call__(self, *args, **kwargs) -> torch.Tensor:
+        return apply_to_collection(self._orig_call(*args, **kwargs), torch.Tensor,
+                                   lambda x: x.to(device=self.device, dtype=self.dtype))
+
+
+class NumpyMetric(Metric):
+    def __init__(self, name: str,
+                 reduce_group: Optional[Any] = torch.distributed.group.WORLD,
+                 reduce_op: Optional[Any] = torch.distributed.ReduceOp.SUM):
+        """
+        Base class for metric implementation operating on numpy arrays.
+        All inputs will be casted to numpy if necessary and all outputs will
+        be casted to tensors if necessary.
+        Already handles DDP sync and input/output conversions
+
+        Args:
+            name: the metric's name
+            reduce_group: the process group for DDP reduces (only needed for DDP training).
+                Defaults to all processes (world)
+            reduce_op: the operation to perform during reduction within DDP (only needed for DDP training).
+                Defaults to sum.
+        """
+        super().__init__(name)
+        self._orig_call = numpy_metric(group=reduce_group,
+                                       reduce_op=reduce_op)(super().__call__)
+
+    def __call__(self, *args, **kwargs) -> torch.Tensor:
+        return apply_to_collection(self._orig_call(*args, **kwargs), torch.Tensor,
+                                   lambda x: x.to(device=self.device, dtype=self.dtype))
```

Comment:
```suggestion
        func_ = lambda x: x.to(device=self.device, dtype=self.dtype)
        return apply_to_collection(self._orig_call(*args, **kwargs), torch.Tensor, func_)
```
easier for debugging